### PR TITLE
Update Vertex AI Gemini provider to use ModelAuthProvider

### PR DESCRIPTION
**Pull request by sberyozkin**
This PR builds on the #694 PR created by @csotiriou and updates Vertex AI Gemini model provider to use ModelAuthProvider. It also adds a demo which depends on the `quarkus-langchain4j-oidc-model-auth-provider` extension which can help to pick up OIDC bearer or authorization code flow tokens.

I'm yet to test it works.

### CODE REVIEW

**Comment by sberyozkin** [comment]
@csotiriou @geoand It all works great, thanks, I'll have a couple of minor suggestions/questions at #694 a bit later, and once #694 is merged, I'll rebase this PR, thanks

**Comment by sberyozkin** [comment]
This is the latest from Gemini:

```
Oh, Java, language of bytecode and beans,
Where objects dance and methods preen.
From applets to servlets, your reach is wide,
A platform for dreams, where code can confide. 

- Gemini
```

PR is ready for review :-), @geoand, @jmartisk, please have a look next week. Jan, since you have tested `secure-fraud-detection`, can you please this demo works ? You'll need to enable Vertex AI in your Google Cloud project, also note this project's id and set in properties, as mentioned in README.

My plan, later, is to add one more, non OIDC authentication into the same demo, and then also use another model, hopefully with Azure OpenAI, so the demo will evolve to show a fairly complex case but also how easily it can be done wit Quarkus LangChain4j and Quarkus security 

Thanks

**Comment by geoand** [comment]
I think this makes sense

**Comment by sberyozkin** [comment]
>When the application has this module, you expect the application to use the user's token, correct?

Right. I'm not sure 100% yet how a situation where 2 remote model providers are used by the application, but only one of them requires a user token, and the `oidc-mode-auth-provider` is loaded, will be handled. I guess we will be able to manage it by either adding the list of models to the `oidc-mode-auth-provider` configuration, or add `oidc-mode-auth-provider` conditionally to the model provider dependencies...

**Comment by geoand** [comment]
Yeah, let's see if this use case makes sense for users before making things more complex

**Comment by sberyozkin** [comment]
The user token will have a user's authorization recorded, for example, when the user is authenticating to Google, the user will be asked to approve a registered application like `Quarkus LangChain4j` to access AI on behalf of the user, if the user does not approve, the authentication will fail

**Comment by sberyozkin** [comment]
Sure, the use case can be generalized as follows in a general OAuth2 way: Quarkus will access downstream services (Genesis model in this case) on behalf of the currently authenticated user only if this user gives a permission, which is quite a mainstream case with SSO..

**Comment by geoand** [comment]
@sberyozkin can you please rebase onto `main`?

**Comment by sberyozkin** [comment]
Sorry @geoand, missed your ping; rebased it

File: samples/secure-vertex-ai-gemini-poem/README.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+quarkus.oidc.client-id=${GOOGLE_CLIENT_ID}
+quarkus.oidc.credentials.secret=${GOOGLE_CLIENT_SECRET}
+quarkus.oidc.authentication.extra-params.scope=https://www.googleapis.com/auth/generative-language.retriever,https://www.googleapis.com/auth/cloud-platform
+quarkus.oidc.authentication.redirect-path=/login
+
+# See https://cloud.google.com/vertex-ai/docs/general/locations
+vertex-ai-region=europe-west2
+
+quarkus.langchain4j.vertexai.gemini.location=https://${vertex-ai-region}-aiplatform.googleapis.com
+quarkus.langchain4j.vertexai.gemini.project-id=${GOOGLE_PROJECT_ID}
```
Comment (jmartisk) [inline samples/secure-vertex-ai-gemini-poem/README.md:27]: Please explain what `GOOGLE_PROJECT_ID` is and how it differs from `GOOGLE_CLIENT_ID`

File: samples/secure-vertex-ai-gemini-poem/README.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+quarkus.oidc.provider=google
+quarkus.oidc.client-id=${GOOGLE_CLIENT_ID}
+quarkus.oidc.credentials.secret=${GOOGLE_CLIENT_SECRET}
+quarkus.oidc.authentication.extra-params.scope=https://www.googleapis.com/auth/generative-language.retriever,https://www.googleapis.com/auth/cloud-platform
+quarkus.oidc.authentication.redirect-path=/login
+
+# See https://cloud.google.com/vertex-ai/docs/general/locations
+vertex-ai-region=europe-west2
+
+quarkus.langchain4j.vertexai.gemini.location=https://${vertex-ai-region}-aiplatform.googleapis.com
```
Comment (jmartisk) [inline samples/secure-vertex-ai-gemini-poem/README.md:26]: In the sample code, you have just
```
quarkus.langchain4j.vertexai.gemini.location=europe-west2
```

(and no property named `vertex-ai-region`), please unify

**Comment by geoand** [comment]
> what is a difference between model-providers/vertex-ai and model-providers/vertex-ai-gemini

They are different models

**Comment by jmartisk** [comment]
> > what is a difference between model-providers/vertex-ai and model-providers/vertex-ai-gemini
> 
> They are different models

That doesn't explain it too much for me - What extension do you use if you want to call, say, Mixtral or Claude through the Google Cloud?

**Comment by geoand** [comment]
You don't currently

File: samples/secure-vertex-ai-gemini-poem/README.md (hunk 1 above)
Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md:28]: @jmartisk I updated the Setup section to explain what Google Project ID is

File: samples/secure-vertex-ai-gemini-poem/README.md (hunk 2 above)
Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md]: Sorry, I though I addressed it but actually not, doing it

Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md]: @jmartisk Done, the full address is not needed by default

**Comment by sberyozkin** [comment]
Interesting, I've got with the latest try:
```
Inspite of the name, 
Java is so much more than a bean.
It's a language that can,
make your wildest dreams a scene. 
Java's an ocean of code,
Where ideas flow and grow. 
It's a haven for the tech-inclined,
A place where innovation can be found.

- Bard
```

As the demo instructs to sign off with the name of the model which produced the poem... `Bard` was the old name for what is now `Gemini`.

@geoand @jmartisk If you happy enough, please merge and I'll start planning to align azure-openai, vertex-ai as well

File: samples/secure-vertex-ai-gemini-poem/README.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+# See https://cloud.google.com/vertex-ai/docs/geeral/locations
+vertex-ai-region=europe-west2
+
+quarkus.langchain4j.vertexai.gemini.location=${vertex-ai-region}
+quarkus.langchain4j.vertexai.gemini.project-id=${GOOGLE_PROJECT_ID}
+```
+
+You must enable Vertex AI API in your Google Cloud project.
+
+Set `GOOGLE_PROJECT_ID` to the id of your Google Cloud project where you have registered the `Quarkus LangChain4j AI` application.
```
Comment (jmartisk) [inline samples/secure-vertex-ai-gemini-poem/README.md:34]: Hold on, isn't it the same thing? `Quarkus LangChain4j AI` is the name of Google Cloud project itself.
Your wording seems to imply that you register applications "under" the project, which I don't think is true (but my knowledge of Google Cloud is very sketchy).

Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md:34]: @jmartisk I'm not sure how else to say it. In Google Cloud, you create a project, and then inside that project you configure things like the application name, etc. For example, my test project name is actually named `Quarkus` - here I have OAuth2 Consent Screen which has a different Application name like `Quarkus LangChain4J AI`, etc, different APIs are enabled.

Shall we we just say `Set`GOOGLE_PROJECT_ID to the id of your Google Cloud project` ?

Comment (jmartisk) [reply samples/secure-vertex-ai-gemini-poem/README.md:34]: Yeah I think "Set GOOGLE_PROJECT_ID to the id of your Google Cloud project" would be clearer :)

Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md:34]: Sure, let me simplify

Comment (sberyozkin) [reply samples/secure-vertex-ai-gemini-poem/README.md]: Done

**Review by jmartisk — APPROVED** [review]

**Pull request merged**